lib/spack/docs/environments.rst

```diff
@@ -368,6 +368,34 @@ from being added again.  At the same time, a spec that already exists in the
 environment, but only as a dependency, will be added to the environment as a
 root spec without the ``--no-add`` option.
 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Developing Packages in a Spack Environment
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``spack develop`` command allows one to develop Spack packages in
+an environment. It requires a spec containing a concrete version, and
+will configure Spack to install the package from local source. By
+default, it will also clone the package to a subdirectory in the
+environment. This package will have a special variant ``develop_path``
+set, and Spack will ensure the package and its dependents are rebuilt
+any time the environment is installed if the package's local source
+code has been modified. Spack ensures that all instances of a
+developed package in the environment are concretized to match the
+version (and other constraints) passed as the spec argument to the
+``spack develop`` command.
+
+For packages with ``git`` attributes, 40-character commit hashes are
+also valid versions. This means that for a package ``foo`` whose
+latest commit is ``abcdef1234abcdef1234abcdef1234abcdef1234``, ``spack
+develop foo@abcdef1234abcdef1234abcdef1234abcdef1234`` will clone the
+latest version of the package, and ``spack install`` will install from
+that source if ``foo`` is in the environment. Further development on
+``foo`` can be tested by reinstalling the environment, and eventually
+committed and pushed to the upstream git repo.
```

Comment:
As a general feedback, I like the feature, but there's a couple concerns, considering this feature seems to be targeted to development, not spack install in general:

- I don't ever think about commits, I work with branches
- I often fork repos when developing them
- When you do development, you make new commits; the "version" number in the spack.yaml however remains as before?

FWIW, in Julia they differentiate between version numbers (using `@`) and git refs (using #), which I think makes a whole lot of sense:

```julia
$ julia -q
(@v1.6) pkg> activate --temp
  Activating new environment at `/tmp/jl_Tl967c/Project.toml`

(jl_Tl967c) pkg> add ArnoldiMethod#master
    Updating git-repo `https://github.com/haampie/ArnoldiMethod.jl.git`
    Updating registry at `~/.julia/registries/General`
   Resolving package versions...
    Updating `/tmp/jl_Tl967c/Project.toml`
  [ec485272] + ArnoldiMethod v0.2.0 `https://github.com/haampie/ArnoldiMethod.jl.git#master`
    Updating `/tmp/jl_Tl967c/Manifest.toml`
  [ec485272] + ArnoldiMethod v0.2.0 `https://github.com/haampie/ArnoldiMethod.jl.git#master`
  [90137ffa] + StaticArrays v1.2.12
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [9a3f8284] + Random
  [9e88b42a] + Serialization
  [2f01184e] + SparseArrays
  [10745b16] + Statistics

(jl_Tl967c) pkg> add ArnoldiMethod@0.2.0
   Resolving package versions...
    Updating `/tmp/jl_Tl967c/Project.toml`
  [ec485272] ~ ArnoldiMethod v0.2.0 `https://github.com/haampie/ArnoldiMethod.jl.git#master` ⇒ v0.2.0
    Updating `/tmp/jl_Tl967c/Manifest.toml`
  [ec485272] ~ ArnoldiMethod v0.2.0 `https://github.com/haampie/ArnoldiMethod.jl.git#master` ⇒ v0.2.0

(jl_Tl967c) pkg> add ArnoldiMethod#1ea8e64648a2d84e4ca973b44c4e08fef0645038
   Resolving package versions...
    Updating `/tmp/jl_Tl967c/Project.toml`
  [ec485272] ~ ArnoldiMethod v0.2.0 ⇒ v0.2.0 `https://github.com/haampie/ArnoldiMethod.jl.git#1ea8e64`
    Updating `/tmp/jl_Tl967c/Manifest.toml`
  [ec485272] ~ ArnoldiMethod v0.2.0 ⇒ v0.2.0 `https://github.com/haampie/ArnoldiMethod.jl.git#1ea8e64`
```
IMHO this is very convenient and also more correct in case of Spack, since spack version "numbers" may be anything, and there are even packages that do `version('develop', branch='master')`...